SYCL/ESIMD/spec_const_redefine_esimd.cpp

```diff
@@ -0,0 +1,149 @@
+
+//==----------- spec_const_redefine_esimd.cpp ------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// The test checks that ESIMD kernels support specialization constants,
+// particularly:
+// - a specialization constant can be redifined and correct new value is used
+//   after redefinition.
+// - the program is JITted only once per a unique set of specialization
+//   constants values.
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/esimd.hpp>
+
+#include <iostream>
+#include <vector>
+
+class SC0;
+class SC1;
+class KernelAAA;
+
+using namespace sycl;
+
+namespace esimd_test {
+// This is the class provided to SYCL runtime by the application to decide
+// on which device to run, or whether to run at all.
+// When selecting a device, SYCL runtime first takes (1) a selector provided by
+// the program or a default one and (2) the set of all available devices. Then
+// it passes each device to the '()' operator of the selector. Device, for
+// which '()' returned the highest number, is selected. If a negative number
+// was returned for all devices, then the selection process will cause an
+// exception.
+class ESIMDSelector : public device_selector {
+  // Require GPU device unless HOST is requested in SYCL_DEVICE_TYPE env
+  virtual int operator()(const device &device) const {
+    if (const char *dev_type = getenv("SYCL_DEVICE_TYPE")) {
+      if (!strcmp(dev_type, "GPU"))
+        return device.is_gpu() ? 1000 : -1;
+      if (!strcmp(dev_type, "HOST"))
+        return device.is_host() ? 1000 : -1;
+      std::cerr << "Supported 'SYCL_DEVICE_TYPE' env var values are 'GPU' and "
+        "'HOST', '"
+        << dev_type << "' is not.\n";
+      return -1;
+    }
+    // If "SYCL_DEVICE_TYPE" not defined, only allow gpu device
+    return device.is_gpu() ? 1000 : -1;
+  }
+};
+inline auto createExceptionHandler() {
+  return [](exception_list l) {
+    for (auto ep : l) {
+      try {
+        std::rethrow_exception(ep);
+      }
+      catch (sycl::exception &e0) {
+        std::cout << "sycl::exception: " << e0.what() << std::endl;
+      }
+      catch (std::exception &e) {
+        std::cout << "std::exception: " << e.what() << std::endl;
+      }
+      catch (...) {
+        std::cout << "generic exception\n";
+      }
+    }
+  };
+}
+} // esimd_test
+
+int val = 0;
+
+// Fetch a value at runtime.
+int get_value() { return val; }
+
+int main(int argc, char **argv) {
+  val = argc;
+
+  queue q(esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler());
+
+  auto dev = q.get_device();
+  std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+  auto ctxt = q.get_context();
+
+  bool passed = true;
+  int x = get_value();
```

Comment:
Do we need to define the global "val" and fetch function "get_value" for this to work?

Reply:
I don't quite understand the objection and proposal here - can you please elaborate?
val and get_value are used to make sure compiler does not optimize/constant propagate.

Reply:
Not objection, just to seek clarification. Since val is assigned argc in the beginning of the function, what happens if this statement is replaced by "int x = val;" or "int x = argc;".

Reply:
I see. Hope my clarification above helps.